src/stores/cpauth/middleware.ts

```diff
@@ -0,0 +1,48 @@
+import CPAuth from 'lib/CPAuth/CPAuth';
+import OAuth2UserImpl from 'lib/OAuth2/OAuth2User';
+import { AnyAction, Middleware } from 'redux';
+import { loadUserError, userExpired } from 'stores/cpauth/actions';
+import { CPAUTH_USER_EXPIRED, CPAUTH_USER_LOAD } from 'stores/cpauth/constants';
+import { getCPAuthUser } from 'stores/cpauth/selectors';
+
+export function cpAuthMiddleware(cpAuth: CPAuth): Middleware {
```

Comment:
This middleware is used to renew user information, if a user is expired.

In general, the user will be renewed automatically, but the flow will fail if there's a race condition between renewal and a separate request. This prevents that.
There is also the case where the token expired when the user was not logged in, so it couldn't be renewed.

This can also prevent actions from getting dispatched, if the user expired, if we want to use this auth method as a primary one.